### ISSUE TITLE
Deployment of OpenMS dll

### ISSUE THREAD

**Issue by timosachsenberg** [OpenMS/OpenMS#4807]
One of the problems we currently face is that we always need to copy parts of the share folder.
It would be great if the user-editable residue, modification databases etc. would be optional.

One way we could achieve this is to parse the XML files only if present.
If they are missing we could fall back to a static version in OpenMS.

We could achieve this by e.g., including the XML file in the share folder in a C++ file so that
we directly parse it from there (e.g., from a string).

This would have some overhead but could make deployment and integration of OpenMS (dll order single tools) much more convenient.

**Comment by cbielow** [comment]
This would need thorough CI testing, because a developer machine usually has SHARE whereas the user binaries would not have them. Big potential for trouble...

Furthermore, if we do not ship share, we also do not ship the examples, RT training models for simulation etc etc... So this creates other usability problems downstream which are very inconvenient for the user.

Technically, including the xml files is certainly possible, but I don't think it's worth the trouble because we'd have a bunch of other issues which would need solving.

**Comment by timosachsenberg** [comment]
I agree that this would need some testing.
I think it would help keep the deployment for the normal OpenMS installers as is (including share, examples etc.).
But for some other things (e.g., real-time processing on the instrument, integration into other GUIs etc. - usually tasks with much more narrow scope) I think it would make sense to compile them in.

**Comment by jpfeuffer** [comment]
I also think that user binaries will have the share most of the time unless they copy them from somewhere (which most of the time crashes anyway because of the dependencies).
However, I would vote to include the DB in a more compact format together with a script that transforms the XML.

**Comment by timosachsenberg** [comment]
would be great to have a constexpr XML parser ;)

**Comment by timosachsenberg** [comment]
To summarize, we have three options so far.
1. always enforce that files in share are parsed (less trouble / error prone but also slower and more difficult to deploy)
2. embed the XML file as a string like in: https://stackoverflow.com/a/47801116/2107184 
3. run a custom script that turns the file into C++ code? Is that what you meant @jpfeuffer ?

**Comment by hroest** [comment]
I think of these options 2 does not make much sense, unless this is done at compile time. We only save time if we actually do (3) and properly model CV terms etc in software. I think proteowizard does something like this (or used to at least). 

We already have code that parses XML as a string, currently that is in the indexed mzML files. Its actually pretty easy

**Comment by hroest** [comment]
Yes, it always has been a bit strange that the library just breaks if it cannot find some text files. Also these are about 200 MB of text files (you can remove the examples) but still its a lot. It also contains libraries such as HMDB2StructMapping.tsv that should not be there in each distribution I think. If these are removed, we talk about ca 8 MB of text that needs to be parsed every time we start OpenMS (unimod + PSI mod already are 3 MB). Also there is some stuff I am not sure we truly need, e.g. brenda.obo ?

PS: this would probably also help deployment for Python / R etc.

**Comment by hroest** [comment]
It looks like we mainly talk about these files (using `File::find`):

```
$ grep -R 'File::find' src/openms | grep -v findExe | grep -v findData | grep -v findDoc |  sed -e 's/.*File::find(\([^)]*\)).*/\1/' | grep '"' | sort | uniq -c | sort -n
      1 "CHEMISTRY/Elements.xml"
      1 "./CHEMISTRY/MissedCleavage.model"
      1 "CHEMISTRY/OMSSA_modification_mapping"
      1 "/CV/psi-mzdata.obo"
      1 "GUISTYLE/qtStyleSheet.qss"
      1 "/MAPPING/mzdata-mapping.xml"
      1 "/MAPPING/mzIdentML-mapping.xml"
      1 "/MAPPING/mzQuantML-mapping_1.0.0-rc2-general.xml"
      1 "/MAPPING/TraML-mapping.xml"
      1 search_name + ".scale"
      1 search_name + ".svm"
      1 "XSL/ConsensusXML.xsl"
      1 "XSL/QcML_report_sheet.xsl"
      2 "/MAPPING/ms-mapping.xml"
      4 "/CV/brenda.obo"
      4 "/CV/goslim_goa.obo"
      4 "/CV/quality.obo"
      5 "/CV/unit.obo"
      7 "/CV/unimod.obo"
     17 "/CV/psi-ms.obo"
```

the question is what other issues we run into if `./share` is missing, for example what does "IDPool" do? I dont know, is it bad if its missing?
